homeassistant/components/prometheus/__init__.py

```diff
@@ -684,6 +684,19 @@ def _handle_switch(self, state: State) -> None:
 
         self._handle_attributes(state)
 
+    def _handle_fan(self, state: State) -> None:
+        metric = self._metric(
+            "fan_state", prometheus_client.Gauge, "State of the fan (0/1)"
+        )
+
+        try:
+            value = self.state_as_number(state)
+            metric.labels(**self._labels(state)).set(value)
+        except ValueError:
+            pass
+
+        self._handle_attributes(state)
```

Comment:
Looking at the set of [supported attributes](https://github.com/home-assistant/core/blob/f09063d70636ae0f87379ebf78de31272fff2974/homeassistant/components/fan/__init__.py#L406-L427) I think only `ATTR_PERCENTAGE` and `ATTR_OSCILLATING` will be automatically converted by `_handle_attributes`, with descriptions that would not be particularly helpful (e.g. "oscillating attribute of fan entity").

Since the number of attributes that `fan` exposes is pretty small, perhaps you would be willing to define metrics for them explicitly, with more helpful descriptions?

I think we could expose the following:
* `fan_speed_percent` (0-100)
* `fan_is_oscillating` (0/1)
* `fan_preset_mode` (enum metric with values 0/1 for each defined preset mode, similar to how we expose `climate_mode`)
* `fan_direction` (enum metric with two labels: `direction="forward"` and `direction="reverse"` and values of 0/1)

In any case, please make sure tests cover additional metrics as well, not just the `fan_state` metric.

Reply:
Thanks for the feedback.  I've added these.  Since direction has only 2 options, I made it a single gauge to reduce cardinality.  Output looks like this:

```
# HELP fan_state State of the fan (0/1)
# TYPE fan_state gauge
fan_state{domain="fan",entity="fan.fan_1",friendly_name="Fan 1"} 1.0
# HELP fan_speed_percent Fan speed percent (0-100)
# TYPE fan_speed_percent gauge
fan_speed_percent{domain="fan",entity="fan.fan_1",friendly_name="Fan 1"} 33.0
# HELP fan_is_oscillating Fan oscillating attribute (0/1)
# TYPE fan_is_oscillating gauge
fan_is_oscillating{domain="fan",entity="fan.fan_1",friendly_name="Fan 1"} 1.0
# HELP fan_preset_mode Fan preset mode
# TYPE fan_preset_mode gauge
fan_preset_mode{domain="fan",entity="fan.fan_1",friendly_name="Fan 1",mode="LO"} 1.0
fan_preset_mode{domain="fan",entity="fan.fan_1",friendly_name="Fan 1",mode="OFF"} 0.0
fan_preset_mode{domain="fan",entity="fan.fan_1",friendly_name="Fan 1",mode="HI"} 0.0
# HELP fan_direction Fan direction (0=forward, 1=reverse)
# TYPE fan_direction gauge
fan_direction{domain="fan",entity="fan.fan_1",friendly_name="Fan 1"} 1.0
# HELP state_change_total The number of state changes
# TYPE state_change_total counter
```